File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.cc
#include <iostream>
#include <vector>
#include <queue>
#include <stack>
using namespace std;


void bfs_search(vector<vector<bool>>& graph, vector<bool>& map,vector<int>& result, int curNode)
{
    for (int i = 1; i < graph.size(); i++)
    {
        if (graph[curNode][i] && !map[i] && curNode != i)
        {
            map[i] = true;
            result.push_back(i);
            bfs_search(graph, map, result, i);
        }
    }
}


void bfs(vector<vector<bool>>& graph, int start)
{
    vector<bool> map(graph.size(), false);
    vector<int> result;

    map[start] = true;
    result.push_back(start);

    bfs_search(graph, map, result, start);
    
    for (int i = 0; i < result.size(); i++)
        cout << result[i] << ' ';

}




void dfs(vector<vector<bool>>& graph, int start)
{
    vector<bool> map(graph.size());
    queue<int> dfsQueue;
    vector<int> result;

    map[start] = true;
    dfsQueue.push(start);
    result.push_back(start);

    while (!dfsQueue.empty())
    {
        int cur = dfsQueue.front();
        dfsQueue.pop();
        for (int i = 1; i < graph.size(); i++)
        {
            if (graph[cur][i] && !map[i] && cur != i)
            {
                dfsQueue.push(i);
                map[i] = true;
                result.push_back(i);
            }
        }
    }

    for (int i = 0; i < result.size(); i++)
        cout << result[i] << ' ';
}




int main(void)
{
    int N, M, start;
    int temp1, temp2;

    cin >> N;
    cin >> M;
    cin >> start;

    vector<vector<bool>> graph(N + 1, vector<bool>(N + 1, false));

    for (int i = 0; i < M; i++)
    {
        cin >> temp1;
        cin >> temp2;
        graph[temp1][temp2] = true;
        graph[temp2][temp1] = true;
    }
    
    
    bfs(graph, start);
    cout << endl;
    dfs(graph, start);
}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.util.*;

public class Main {

	static boolean[][] graph;
	
	
	static boolean[] isVisit;
    static ArrayList<Integer> result;
    
    
	static void dfs(int curNode){
	    for (int i = 0; i < graph.length; i++){
	        if (graph[curNode][i] && !isVisit[i]){
	            isVisit[i] = true;
	            result.add(i);
	            dfs(i);
	        }
	    }
	}


	static void bfs(int start){
		
	    Queue<Integer> dfsQueue = new LinkedList<Integer>();;
	    
	    isVisit[start] = true;
	    dfsQueue.add(start);
	    result.add(start);

	    while (!dfsQueue.isEmpty()){
	        int cur = dfsQueue.poll();
	        
	        for (int i = 0; i < graph.length; i++){
	            if (graph[cur][i] && !isVisit[i]){
	                dfsQueue.add(i);
	                isVisit[i] = true;
	                result.add(i);
	            }
	        }
	    }
	}


	public static void main(String args[]) {
	    
		Scanner scanner = new Scanner(System.in);
		

	    int N = scanner.nextInt();
	    int M = scanner.nextInt();
	    int start = scanner.nextInt();

	    graph = new boolean[N][N];

	    for (int i = 0; i < M; i++)
	    {
	        int temp1 = scanner.nextInt();
	        int temp2 = scanner.nextInt();
	        graph[temp1-1][temp2-1] = true;
	        graph[temp2-1][temp1-1] = true;
	    }
	    
	    
	    isVisit = new boolean[graph.length];
	    result = new ArrayList<Integer>();;
	    
	    
	    isVisit[start-1] = true;
	    result.add(start-1);
	    dfs(start-1);
	    for(int n : result) 
	    	System.out.print(n+1 + " ");
	    
	   
	    System.out.println();

	    
	    isVisit = new boolean[graph.length];
	    result = new ArrayList<Integer>();;
	    
	    bfs(start-1);
	    for(int n : result) 
	    	System.out.print(n+1 + " ");
	    
	    
	    
	}
}